kernel/of/fdt_mem.c

```diff
@@ -0,0 +1,192 @@
+        }
+        else {
+            // Trim from above
+            end = resv_start;
+            *size_ptr = (size_t)(end - start);
+            return false;
+        }
+    }
+
+    // Should be unreachable
+}
+
+// Try to free this region (return the number of bytes successfully freed)
+static size_t
+try_boot_free_region(struct fdt* fdt, void* start, size_t size)
+{
+    // Should be unnecessary (but it's worth checking in case someone messes up)
+    if (size == 0) {
+        return (size_t)0;
+    }
+
+    // Buffers for a second region if we get split into two
+    void* split_start;
+    size_t split_size;
+
+    // Check if this is a region before the end of the kernel (protects the kernel and
+    // firmware which is before it)
+    extern int __kernel_end[];
+    if (start < (void*)__kernel_end) {
+        if ((void*)(start + size) <= (void*)__kernel_end) {
+            // Entirely before the kernel
+            return (size_t)0;
+        }
+        else {
+            // Need to trim before
+            size = (size_t)((start + size) - (void*)__kernel_end);
+            start = (void*)__kernel_end;
+        }
+    }
+    else {
+        // No addresses lower than the end of the kernel (so no overlap)
+    }
+
+    // Handle the reserve entries in the FDT
+    struct fdt_reserve_entry* entry = fdt_reserve_entry_begin(fdt);
+    while (entry != NULL) {
+        void* resv_start = fdt_reserve_entry_address(entry);
+        size_t resv_size = fdt_reserve_entry_size(entry);
+
+        if (trim_or_split_around_reserved(
+                &start, &size, &split_start, &split_size, resv_start, resv_size
+            )) {
+            // We split into 2 regions
+            return try_boot_free_region(fdt, start, size)
+                   + try_boot_free_region(fdt, split_start, split_size);
+        }
+        else {
+            if (size == 0) {
+                // We got fully trimmed
+                return (size_t)0;
+            }
+            else {
+                // We're still one contiguous region (continue)
+            }
+        }
+
+        entry = fdt_next_reserve_entry(fdt, entry);
+    }
+
+    // Make sure we don't free the FDT blob itself
+    void* fdt_resv_start = (void*)fdt;
+    size_t fdt_resv_size = fdt_size(fdt);
+
+    if (trim_or_split_around_reserved(
+            &start, &size, &split_start, &split_size, fdt_resv_start, fdt_resv_size
+        )) {
+        // We split into 2 regions
+        return try_boot_free_region(fdt, start, size)
+               + try_boot_free_region(fdt, split_start, split_size);
+    }
+    else {
+        if (size == 0) {
+            // We got fully trimmed
+            return (size_t)0;
+        }
+        else {
+            // We're still one contiguous region (continue)
+        }
+    }
+
+    // We made it through (possibly shrunk)
+    if (boot_free(start, size) != KERR_SUCCESS) {
+        // We couldn't free for some reason
+        return (size_t)0;
```

Comment:
```suggestion
        return 0;
```

Reply:
I'm not changing this (or any of the other similar situations), it doesn't hurt anything and is a good reminder to check exact what your return type is and why you're returning the `0`.

Reply:
Because we can't always be perfect

I said it's `good` not needed. Doing it always would be ideal, doing it sometimes is a nice thing, and doing it never would be sad but not a big deal.

Reply:
`sed -i 's/return 0/return (size_t)0/g' <files>`

Reply:
That changed nothing, there's no `return 0`s in that file